### DISCUSSION TITLE
Crop a fixed-size region surrounding tumor centroid

### DISCUSSION THREAD

**Discussion by mattwarkentin** [Project-MONAI/MONAI#3040] [General]
Hi,

I am trying to perform a transform that, for each subject, will crop a volume of fixed size centred around a coordinate. The coordinate corresponds to the center of the tumor in `'mask'`. This is a pretty common task. However, the coordinate varies for each subject because it corresponds to the location of a tumor. It seems like when you instantiate the `SpatialCrop`/`SpatialCropd` transforms you need to input a `roi_center`, but shouldn't this support centers that are patient-specific?

Something like this seems like it would make sense:
```py
sub = {
  'img': 'image.nrrd',
  'mask': 'mask.nrrd',
  'coord': (100, 110, 75)
}

SpatialCropd(['image'], roi_center_key='coord', roi_size=(50, 50, 50))
```

The code below is as close as I've gotten to what I desire so far, but it firsts crops based on the masks bounding box, then resizes to the desired dimension `(50, 50, 50)`. But if the mask is smaller than 50x50x50, the final image centers on the immediate nodule bounding box and is padded with black space, not ideal.
```py
keys = ['img', 'mask']

trans = Compose([
  LoadImaged(keys),
  AddChanneld(keys),
  CropForegroundd(keys, source_key='mask'),
  ResizeWithPadOrCropd(keys, spatial_size=(50, 50, 50)),
  ToTensord(keys)
])

sub_t = trans(sub)
```
What I want is a crop of `'img'` that is 50x50x50 in size and is centred on the tumor.

**Comment by rijobro** [comment]
Hi @mattwarkentin , I think the easiest thing would be to create your own transformation that inherits from `MapTransform`. During the `__call__`, you would extract the coordinates from one of the other keys in your data.

Something like this (untested):
```python
class DynamicCrop(MapTransform):
    def __init__(self, keys, coord_key):
        super().__init__(keys)
        self.coord_key = coord_key
    def __call__(self, data):
        d = dict(data)
        coord = d[self.coord_key]
        cropper = SpatialCrop(...)  # use the coordinates and your desired size to create a cropper
        for k in self.keys:
            d[k] = cropper(d[k])
```

**Reply by mattwarkentin** [reply]
Great suggestion! Much appreciated. Still learning the flexibility of MONAI.

**Comment by mattwarkentin** [comment]
For posterity, this function works as described in the original post and will ALWAYS return an object of the desired size...
```py
class DynamicCropOrPad(monai.transforms.MapTransform):
    def __init__(self, keys, coord_key, roi_size):
        super().__init__(keys)
        self.coord_key = coord_key
        self.roi_size = roi_size
    def __call__(self, data):
        d = dict(data)
        coord = d[self.coord_key]
        cropper = SpatialCrop(roi_center=coord, roi_size=self.roi_size)
        resizer = ResizeWithPadOrCrop(spatial_size=self.roi_size)
        for k in self.keys:
            d[k] = cropper(d[k])
            if (tuple(d[k].shape) != self.roi_size):
              d[k] = resizer(d[k])
        return(d)
```

**Reply by rijobro** [reply]
Great, glad you got it working!

**Reply by LalehSeyyed** [reply]
Hi @mattwarkentin, I am wondering how do you define coord_key. as you mentioned the coord_key varies per patient based on the location of the tumor. Do you feed the location of the tumor from your dataset or do you use a code to automatically locate the coord_key from your segmentation mask labels?

**Reply by rijobro** [reply]
Although the tumour centre varies between patients, the key used to store that information in the data dictionary can remain constant.

For example, your data might have the following structure:

```python
data = {
    "image": torch.Tensor(...),
    "label": torch.Tensor(...),
    "tumor_center": torch.Tensor(...),  # coordinates that point to tumour centre. can change between patients, but key is always the same.
}
```

Edit - sorry I completely misread your question @LalehSeyyed! There are lots of way of generating the coordinates, using another network, for example. I'll let @mattwarkentin let you know how he does it in his specific case.

**Reply by mattwarkentin** [reply]
Hi @LalehSeyyed. In my case, the coordinates for the centre of the tumour were reported by radiologists who had previously reviewed the scans. But as @rijobro says, the coordinate value could be obtained using any method really. If you have a segmentation mask, you could also probably obtain coordinates by finding the centre of a label but it might take some extra code to get that information.
